urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/lp/solver/APRIntervalHybridCohoSolver.java

```diff
@@ -0,0 +1,416 @@
+	//2)Use history to check the less favorable one. Check cohoDual feasiblity for basis if it's not clearly feasible to remove the infeasible basis.
+	protected LPBasis pivot(LPBasis currBasis, Set<LPBasis> history) throws UnboundedLPError, OptInHistoryException{
+		try{//try to use interval
+			LPBasis nextBasis = pivot(currBasis,0,false);//From the begining, find a possible favorable basis.
+			if(nextBasis==null){//optimal
+				return null;
+			}else{
+				//If dualStatus is not clear feasible, it might be the case.(check it) 
+				//1)currBasis is optimal 2) nextBasis is infeasible or less favorable.
+				//The dualStatus might be not clear feasible if currBasis is not optimal. In this case, the check is false.
+				if(nextBasis.dualStatus()!=LPBasis.BasisStatus.FEASIBLE){
+					//not a clear favorable basis found, Compute the feasiblity of nextBasis use apr.
+					//If currBasis is optimal. Interval might introduce non-invertiable basis thus cause a SingularMatrixException and call apr then.
+					//NOTE: I override the cohoDualFeasile, it will not throw LPError if the basis is not invertiable. The status is infeasible now. 
+					//In this case, we don't need to try the apr.
+					cohoDualFeasible(nextBasis);
+					//nextBasis is infeasible, then currBasis must be optimal. Otherwise, it is impossible that there is only one infeasible branch.
+					if(nextBasis.dualStatus()==LPBasis.BasisStatus.INFEASIBLE)
+						return null;
+					
+					//nextBasis is feasible, but might be less favorable. Then the optimal basis lead to 
+					//some feasible basis which might never shouwn before.
+					//But at the end, it will introduce a loop and then we can find the optimal one from history. 
+					if(history.contains(nextBasis)){//see the equals function of LPBasis, it's independent of status
+						throw new OptInHistoryException();
+					}else{
+						//the basis is really feasible and not appeared before, try it and continue
+					}
+				}
+				return nextBasis;
+			}
+		}catch(LPError e){//CONSIDER: change to Exception? We have considered the SingularMatrixException and Divided by zero.
+			if(e instanceof UnboundedLPError)
+				throw e;
+			//throw SingularMatrixException. call apr pivot
+		}		
+		//use apr
+		return super.pivot(currBasis);
+	}
+	
+	//
+	//Find the next possible more favorable basis. Return null if there is no better one. Throw LPError if more than one next basis found. 
+    //The result might be less favorable or infeasible when the currBasis is optimal. It's indicated by the dual status as POSSIBLEFEASIBLE.
+	//
+	//What about the Singular basis? 
+	//Most singular basis are introduced by BIGM variable. We use the same method (see interval solver) to remove it as possible.
+	//If it's optimal, return as result. otherwise, throw Exception and use APR in the next pivot
+	//
+	private LPBasis pivot(LPBasis currBasis, int start, boolean findClear) throws UnboundedLPError{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.